plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -149,6 +178,68 @@ public StatisticsFile writeStatisticsFile(
+            long snapshotId,
+            FileIO fileIO,
+            StatsUpdateMode updateMode,
+            CollectedStatistics collectedStatistics)
+    {
+        if (updateMode == INCREMENTAL_UPDATE) {
+            Snapshot snapshot = table.snapshot(snapshotId);
+            checkState(snapshot != null, "No snapshot information for snapshotId %s in table %s", snapshotId, table);
+            if (snapshot.parentId() == null || !maySnapshotHaveData(table, snapshot.parentId(), fileIO)) {
+                // No previous snapshot, or previous snapshot empty
+                updateMode = REPLACE;
+            }
+        }
+
+        return switch (updateMode) {
+            case REPLACE -> collectedStatistics;
+            case INCREMENTAL_UPDATE -> {
+                Map<Integer, CompactSketch> collectedNdvSketches = collectedStatistics.ndvSketches();
+                ImmutableMap.Builder<Integer, CompactSketch> ndvSketches = ImmutableMap.builder();
+
+                Set<Integer> pendingPreviousNdvSketches = new HashSet<>(collectedNdvSketches.keySet());
+                Iterator<StatisticsFile> statisticsFiles = walkStatisticsFiles(table, snapshotId);
+                while (!pendingPreviousNdvSketches.isEmpty() && statisticsFiles.hasNext()) {
+                    StatisticsFile statisticsFile = statisticsFiles.next();
+
+                    boolean hasUsefulData = statisticsFile.blobMetadata().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.fields().size() == 1)
+                            .anyMatch(blobMetadata -> pendingPreviousNdvSketches.contains(getOnlyElement(blobMetadata.fields())));
+
+                    if (hasUsefulData) {
+                        try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
+                                .withFileSize(statisticsFile.fileSizeInBytes())
+                                .withFooterSize(statisticsFile.fileFooterSizeInBytes())
+                                .build()) {
+                            List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
+                                    .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
+                                    .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
+                                    .filter(blobMetadata -> pendingPreviousNdvSketches.contains(getOnlyElement(blobMetadata.inputFields())))
+                                    .collect(toImmutableList());
+                            for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
+                                Integer fieldId = getOnlyElement(read.first().inputFields());
+                                checkState(pendingPreviousNdvSketches.remove(fieldId), "Unwanted read of stats for field %s", fieldId);
```

Comment:
This doesn't seem like a safe assumption, unless I'm missing something.

You're walking the snapshot history back until you find all the columns you're looking for so you may find duplicates for the columns you did find already.

Reply:
i think it is, so will ignore this for now

Reply:
You are right, I missed that `filter` line